### PR TITLE
Better error handling, suggester, and force async in command signature

### PR DESCRIPTION
**Pull request by Defman**

### CODE REVIEW

File: benches/dispatcher.rs [hunk 1]
```diff
@@ -3,16 +3,31 @@ use lieutenant::{command, CommandDispatcher, Context};
...
+    fn from(_: std::convert::Infallible) -> Self {
+        panic!("Cannot fail")
+    }
+}
 
 fn single_command(c: &mut Criterion) {
     struct State;
     impl Context for State {
-        type Error = Error;
+        type Err = Error;
```
Comment (caelunshun) [inline benches/dispatcher.rs]: This should still be `Error`, not `Err`, to remain consistent with the standard library

File: macros/src/lib.rs [hunk 2]
```diff
@@ -43,6 +43,10 @@ pub fn command(
         ),
     };
 
+    if let None = input.sig.asyncness {
```
Comment (caelunshun) [inline macros/src/lib.rs]: Could be `if input.sig.asyncness.is_none()`

File: macros/src/lib.rs [hunk 3]
```diff
@@ -43,6 +43,10 @@ pub fn command(
...
+        abort_call_site!("command must be async fn");
```
Comment (caelunshun) [inline macros/src/lib.rs]: I think this should be `emit_call_site`, since this isn't a fatal error.

File: src/parser.rs [hunk 4]
```diff
@@ -37,21 +39,37 @@ pub trait ArgumentChecker<C>: Any + Send + Sync + 'static {
...
     where
         Self: Sized;
 }
 
-pub trait ArgumentKind<C: Context>: Sized + Send + Sync {
+pub trait ArgumentSuggester<C>
+where
+    C: Context,
+{
+    fn suggestions<'a, 'b, 'c>(&'a self, _ctx: &'b C, _input: &'c str) -> FutureBox<'c, Vec<String>>;
```
Comment (caelunshun) [inline src/parser.rs]: Could the return value be `Cow<'static, str>` instead, to reduce allocations in trivial cases?

File: benches/dispatcher.rs (hunk 1 above)
Comment (Defman) [reply benches/dispatcher.rs]: That's why I changed it to `Err`, `FromStr::Err`, or is the standard library just not consistent?

Comment (caelunshun) [reply benches/dispatcher.rs]: The naming of `FromStr::Err` is widely regarded as a mistake in the standard library, but they can't change it due to backwards compatibility. All other error associated types in `std` are called `Error`.

**Comment by caelunshun** [comment]
Thanks!

**Pull request merged**